tools/graph_loading/construct_graph.py

```diff
@@ -0,0 +1,680 @@
+                "transform":    {"name": "<operator name>", ...}
+            },
+        ],
+        "labels":       [
+            {
+                "label_col":    "<column name>",
+                "task_type":    "<task type: e.g., classification>",
+                "split_type":   [0.8, 0.2, 0.0],
+                "custom_train": "<the file with node IDs in the train set>",
+                "custom_valid": "<the file with node IDs in the validation set>",
+                "custom_test":  "<the file with node IDs in the test set>",
+            },
+        ],
+    }
+
+    Parameters
+    ----------
+    process_confs: list of dicts
+        The configurations to process node data.
+    remap_id: bool
+        Whether or not to remap node IDs
+
+    Returns
+    -------
+    dict: node ID map
+    dict: node features.
+    """
+    node_data = {}
+    node_id_map = {}
+    for process_conf in process_confs:
+        assert 'node_id_col' in process_conf, \
+                "'node_id_col' must be defined for a node type."
+        node_id_col = process_conf['node_id_col']
+        assert 'node_type' in process_conf, \
+                "'node_type' must be defined for a node type"
+        node_type = process_conf['node_type']
+        assert 'format' in process_conf, \
+                "'format' must be defined for a node type"
+        read_file = parse_file_format(process_conf['format'])
+        assert 'files' in process_conf, \
+                "'files' must be defined for a node type"
+        in_files = get_in_files(process_conf['files'])
+        feat_ops = parse_feat_ops(process_conf['features']) \
+                if 'features' in process_conf else None
+        label_conf = process_conf['labels'] if 'labels' in process_conf else None
+        q = []
+        manager = multiprocessing.Manager()
+        return_dict = manager.dict()
+        for i, in_file in enumerate(in_files):
+            p = Process(target=parse_node_data, args=(i, in_file, feat_ops, node_id_col,
+                                                      label_conf, read_file, return_dict))
+            p.start()
+            q.append(p)
+            wait_process(q, num_processes)
+        wait_all(q)
+
+        type_node_id_map = [None] * len(return_dict)
+        type_node_data = {}
+        for i, (node_ids, data) in return_dict.items():
+            for feat_name in data:
+                if feat_name not in type_node_data:
+                    type_node_data[feat_name] = [None] * len(return_dict)
+                type_node_data[feat_name][i] = data[feat_name]
+            type_node_id_map[i] = node_ids
+
+        assert type_node_id_map[0] is not None
+        type_node_id_map = np.concatenate(type_node_id_map)
+        # We don't need to create ID map if the node IDs are integers,
+        # all node Ids are in sequence start from 0 and
+        # the user doesn't force to remap node IDs.
+        if np.issubdtype(type_node_id_map.dtype, np.integer) \
```

Comment:
Are you assuming the node_id_maps are ordered in type_node_id_map? Is manager.dict() an sorted dict (key-val pairs are stored according to the order of keys) or not?

Reply:
i don't understand what you mean. i don't think we need to assume it's sorted.